pkg/boot/uefi/uefi.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package uefi
+
+import (
+	"bytes"
+	"debug/pe"
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/u-root/u-root/pkg/acpi"
+	"github.com/u-root/u-root/pkg/boot"
+	"github.com/u-root/u-root/pkg/boot/kexec"
+	"github.com/u-root/u-root/pkg/smbios"
+)
+
+const fvEntryImageOffset int64 = 0xA0
+
+var bootExecute = boot.Execute
+var kexecLoad = kexec.Load
+var kexecParseMemoryMap = kexec.ParseMemoryMap
+var getRSDP = acpi.GetRSDP
+
+func getSMBIOSBaseInternal() (int64, int64, error) {
+	base, size, err := smbios.GetSMBIOSBaseEFI()
+	if err != nil {
+		base, size, err = smbios.GetSMBIOSBaseLegacy()
+		if err != nil {
+			return 0, 0, err
+		}
+	}
+	return base, size, nil
+}
+
+var getSMBIOSBase = getSMBIOSBaseInternal
+
+// Serial port setting in Linuxboot.h
+const (
+	SerialPortTypeIO   = 1
+	SerialPortTypeMMIO = 2
+)
+
+// SerialPortConfig defines debug port configuration
+type SerialPortConfig struct {
+	Type        uint32
+	BaseAddr    uint32
+	Baud        uint32
+	RegWidth    uint32
+	InputHertz  uint32
+	UartPciAddr uint32
+}
+
+type payloadConfig struct {
+	AcpiBase            uint64
+	AcpiSize            uint64
+	SmbiosBase          uint64
+	SmbiosSize          uint64
+	SerialConfig        SerialPortConfig
+	NumMemoryMapEntries uint32
+}
+
+// FvImage is a structure for loading a firmware volume
+type FvImage struct {
+	name         string
+	mem          kexec.Memory
+	entryAddress uintptr
+	ImageBase    uintptr
+	SerialConfig SerialPortConfig
+}
+
+// PEImage is a structure for loading entry PE image from a firmware volume
+type PEImage struct {
+	file *os.File
+}
+
+// ReadAt of PEImage will add fvEntryImageOffset to regular ReadAt's offset
```

Comment:
You can probably just use an io.SectionReader().

Reply:
Done